utils/gcp/object_storage.cc

```diff
@@ -0,0 +1,820 @@
+{}
+
+utils::gcp::storage::client::impl::impl(std::string_view endpoint, std::optional<google_credentials> c, shared_ptr<seastar::tls::certificate_credentials> certs)
+    : impl(rest::parse_simple_uri(endpoint.empty() ? STORAGE_APIS_URI : endpoint), std::move(c), std::move(certs))
+{}
+
+using status_type = seastar::http::reply::status_type;
+
+/**
+ * Performs a REST post/put/get with credential refresh/retry.
+ */
+future<> 
+utils::gcp::storage::client::impl::send_with_retry(const std::string& path, const std::string& scope, body_variant body, std::string_view content_type, handler_func_ex handler, httpclient::method_type op, key_values headers) {
+    int retries = 0;
+
+    for (;;) {
+        rest::request_wrapper req(_endpoint);
+        req.target(path);
+        req.method(op);
+
+        if (_credentials) {
+            try {
+                co_await _credentials->refresh(scope, &storage_scope_implies, _certs);
+                req.add_header(utils::gcp::AUTHORIZATION, format_bearer(_credentials->token));
+            } catch (...) {
+                std::throw_with_nested(permission_error("Error refreshing credentials"));
+            }
+        }
+
+        for (auto& [k,v] : headers) {
+            req.add_header(k, v);
+        }
+
+        if (!content_type.empty()) {
+            req.add_header(httpclient::CONTENT_TYPE_HEADER, content_type);
+        }
+
+        std::visit(overloaded_functor {
+            [&](const std::string& s) { req.content(s); },
+            [&](const writer_and_size& ws) { req.content(ws.first, ws.second); }
+        }, body);
+
+        // GCP storage requires this even if content is empty
+        req.add_header("Content-Length", std::to_string(req.request().content_length));
+
+        gcp_storage.trace("Sending: {}", req.request());
+
+        try {
+            co_await rest::simple_send(_client, req, [&handler](const seastar::http::reply& res, seastar::input_stream<char>& in) -> future<> {
+                gcp_storage.trace("Result: {}", res);
+                if (res._status == status_type::unauthorized) {
+                    throw httpd::unexpected_status_error(res._status);
+                }
+                co_await handler(res, in);
+            });
+            break;
+        } catch (httpd::unexpected_status_error& e) {
+            gcp_storage.debug("{}: Got unexpected response: {}", _endpoint, e.status());
+            if (e.status() == status_type::unauthorized && retries++ < 3) {
+                // refresh access token and retry.
+                continue;
+            }
+            if (e.status() == status_type::unauthorized) {
+                std::throw_with_nested(permission_error(std::string(_endpoint)));
+            }
+            throw;
+        } catch (...) {
+            // network, whatnot. maybe add retries here as well, but should really 
+            // be on seastar level
```

Comment:
Yes. https://github.com/scylladb/seastar/issues/2803

Reply:
They use the same approach as AWS does, what does http code 429 mean here? In fact it is 5 different erroneous states. Some are retryable, some not, in any case, you will be unable to determine the root cause of the failure once one http code has multiple meanings, so how are you going to handle it in response callbacks (I guess this is what you mean by "calling routines"), each time `switch...case` http codes and parsing erroneous responses?

Reply:
We don't really have to handle error codes very specificly. Ok and created, and partial depending on context. 5xx should be retried (yes, should eventually move to shared code maybe). As for 429 etc; Not sure we need that level of specificity. Maybe adding extracting error messages, yes, but that is not retryability, just info for throwing.

Reply:
It is not about 429 or any other http code we can get from the server. My point is simple, once you get a 4XX code, you log it and then, in your opinion, the engineer investigating failure and searching through logs will understand what exactly went wrong on GSC side when encountering such a log line?
For example, Forbidded has 15 reasons to be sent to the client
https://cloud.google.com/storage/docs/xml-api/reference-status#403%E2%80%94forbidden